lib/spack/spack/preferred_packages.py

```diff
@@ -154,7 +154,10 @@ def spec_has_preferred_provider(self, pkgname, provider_str):
 
     def spec_preferred_variants(self, pkgname):
         """Return a VariantMap of preferred variants and their values"""
-        variants = self.preferred.get(pkgname, {}).get('variants', '')
+        for pkg in (pkgname, 'all'):
+            variants = self.preferred.get(pkg, {}).get('variants', '')
```

Comment:
@suraia: This will fail if the package doesn't support the particular global variant.  It sounds like it worked for you because you added a variant to `m4`, but it would be nice to prefer `all: +debug` even when some packages don't have debug.  I'll put how to fix it below.